nifi-registry-core/nifi-registry-framework/src/main/java/org/apache/nifi/registry/service/extension/StandardExtensionService.java

```diff
@@ -0,0 +1,589 @@
+
+            // persist the content of the bundle to the persistence provider
+            final ExtensionBundleContext context = new StandardExtensionBundleContext.Builder()
+                    .bundleType(getProviderBundleType(bundleType))
+                    .bucketId(existingBucket.getId())
+                    .bucketName(existingBucket.getName())
+                    .bundleId(extensionBundle.getId())
+                    .bundleGroupId(extensionBundle.getGroupId())
+                    .bundleArtifactId(extensionBundle.getArtifactId())
+                    .bundleVersion(versionMetadata.getVersion())
+                    .author(versionMetadata.getAuthor())
+                    .timestamp(versionMetadata.getTimestamp())
+                    .build();
+
+            try (final InputStream in = new FileInputStream(extensionWorkingFile);
+                 final InputStream bufIn = new BufferedInputStream(in)) {
+                bundlePersistenceProvider.saveBundleVersion(context, bufIn);
+                LOGGER.debug("Bundle saved to persistence provider - '{}' - '{}' - '{}'",
+                        new Object[]{groupId, artifactId, version});
+            }
+
+            // get the updated extension bundle so it contains the correct version count
+            final ExtensionBundleEntity updatedBundle = metadataService.getExtensionBundle(bucketIdentifier, groupId, artifactId);
+
+            // create the full ExtensionBundleVersion instance to return
+            final ExtensionBundleVersion extensionBundleVersion = new ExtensionBundleVersion();
+            extensionBundleVersion.setVersionMetadata(versionMetadata);
+            extensionBundleVersion.setExtensionBundle(DataModelMapper.map(existingBucket, updatedBundle));
+            extensionBundleVersion.setBucket(DataModelMapper.map(existingBucket));
+            return extensionBundleVersion;
+
+        } finally {
+            if (extensionWorkingFile.exists()) {
+                try {
+                    extensionWorkingFile.delete();
+                } catch (Exception e) {
+                    LOGGER.warn("Error removing temporary extension bundle file at {}",
+                            new Object[]{extensionWorkingFile.getAbsolutePath()});
+                }
+            }
+        }
+    }
+
+    private ExtensionBundleEntity getOrCreateExtensionBundle(final String bucketId, final String groupId,
+                                                             final String artifactId, final ExtensionBundleType bundleType) {
+        ExtensionBundleEntity existingBundleEntity = metadataService.getExtensionBundle(bucketId, groupId, artifactId);
+        if (existingBundleEntity == null) {
+            final ExtensionBundle bundle = new ExtensionBundle();
+            bundle.setIdentifier(UUID.randomUUID().toString());
+            bundle.setBucketIdentifier(bucketId);
+            bundle.setName(groupId + " - " + artifactId);
```

Comment:
Don't know how others feel about this, but personally I would have a slight preference for using the Maven coordinate style here for the inferred bundle name, i.e.:

    bundle.setName(groupId + ":" + artifactId);

Reply:
Makes sense, I'm on-board with that